app/src/main/java/mozilla/lockbox/store/FingerprintStore.kt

```diff
@@ -0,0 +1,201 @@
+package mozilla.lockbox.store
+
+import android.hardware.fingerprint.FingerprintManager
+import android.os.CancellationSignal
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyPermanentlyInvalidatedException
+import android.security.keystore.KeyProperties
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.rxkotlin.addTo
+import io.reactivex.subjects.ReplaySubject
+import mozilla.lockbox.action.AuthenticationAction
+import mozilla.lockbox.extensions.filterByType
+import mozilla.lockbox.flux.Dispatcher
+import java.io.IOException
+import java.security.InvalidAlgorithmParameterException
+import java.security.InvalidKeyException
+import java.security.KeyStore
+import java.security.KeyStoreException
+import java.security.NoSuchAlgorithmException
+import java.security.NoSuchProviderException
+import java.security.UnrecoverableKeyException
+import java.security.cert.CertificateException
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.NoSuchPaddingException
+import javax.crypto.SecretKey
+
+open class FingerprintStore(
+    val dispatcher: Dispatcher = Dispatcher.shared
+) {
+    internal val compositeDisposable = CompositeDisposable()
+    private lateinit var fingerprintManager: FingerprintManager
+
+    private lateinit var keyStore: KeyStore
+    private lateinit var keyGenerator: KeyGenerator
+    private var cancellationSignal: CancellationSignal? = null
+    private var selfCancelled = false
+
+    val authState: ReplaySubject<AuthenticationState> = ReplaySubject.create()
+
+    companion object {
+        val shared = FingerprintStore()
+        private const val ANDROID_KEY_STORE = "AndroidKeyStore"
+        private const val DEFAULT_KEY = "DEFAULT_KEY"
+    }
+
+    sealed class AuthenticationState {
+        object Succeeded : AuthenticationState()
+        object Failed : AuthenticationState()
+        data class Error(val error: String?) : AuthenticationState()
+    }
+
+    init {
+        dispatcher.register.filterByType(AuthenticationAction::class.java)
+            .subscribe {
+                when (it) {
+                    is AuthenticationAction.StartListening -> initFingerprint()
+                    is AuthenticationAction.StopListening -> stopListening()
+                }
+            }
+            .addTo(compositeDisposable)
+    }
+
+    fun apply(manager: FingerprintManager) {
+        fingerprintManager = manager
+    }
+
+    fun isFingerprintAuthAvailable(): Boolean {
```

Comment:
We may want this to eventually be an `Observable<Boolean>` so that we can include the user-setting for fingerprint usage as well.

Reply:
I'm expecting this to be addressed in #31

Reply:
👍